{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5141b2c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/misaell2/.conda/envs/besca/lib/python3.9/site-packages/besca/datasets/_datasets.py:8: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.10.3 anndata==0.10.8 umap==0.5.9.post2 numpy==1.26.4 scipy==1.13.1 pandas==2.3.3 scikit-learn==1.6.1 statsmodels==0.14.5 igraph==1.0.0 leidenalg==0.11.0 pynndescent==0.5.13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/misaell2/.conda/envs/besca/lib/python3.9/site-packages/umap/__init__.py:9: ImportWarning:\n",
      "\n",
      "Tensorflow not installed; ParametricUMAP will be unavailable\n",
      "\n",
      "/home/misaell2/.conda/envs/besca/lib/python3.9/site-packages/scanpy/_settings.py:488: DeprecationWarning:\n",
      "\n",
      "`set_matplotlib_formats` is deprecated since IPython 7.23, directly use `matplotlib_inline.backend_inline.set_matplotlib_formats()`\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import besca as bc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import sparse, io\n",
    "import scipy.sparse as sp\n",
    "import os\n",
    "import re\n",
    "import time\n",
    "import logging\n",
    "import seaborn as sns\n",
    "import h5py\n",
    "sc.logging.print_header()\n",
    "\n",
    "# for standard processing, set verbosity to minimum\n",
    "sc.settings.verbosity = 0  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.settings.set_figure_params(dpi=80)\n",
    "version = '2.8'\n",
    "start0 = time.time()\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "702b56d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define standardized filepaths based on above input\n",
    "root_path = os.getcwd()\n",
    "bescapath_full = os.path.dirname(bc.__file__)\n",
    "bescapath = os.path.split(bescapath_full)[0]\n",
    "\n",
    "analysis_name = 'all_merged_bbknn'\n",
    "\n",
    "# Choose the clustering to base the annotation on.\n",
    "# leiden (RNA), CLR (CITE-seq), DSB (CITE-seq)\n",
    "clusters='leiden' \n",
    "\n",
    "batch_to_correct = 'ID'\n",
    "\n",
    "results_folder = os.path.join(root_path, 'analyzed', analysis_name)\n",
    "\n",
    "# Input h5ad file\n",
    "input_annotated = os.path.join(root_path, 'annotated.all_merged.h5ad')\n",
    "input_raw = os.path.join(root_path, 'raw.all_merged.h5ad')\n",
    "\n",
    "# Output h5ad file\n",
    "results_file = os.path.join(results_folder, analysis_name + '.annotated.clustered.h5ad')\n",
    "\n",
    "#define output directories\n",
    "outdir_figures = figdir = os.path.join(results_folder, 'figures')\n",
    "outdir_results = os.path.join(results_folder, 'results')\n",
    "outdir_export_data = os.path.join(results_folder, 'outfiles')\n",
    "\n",
    "sc.settings.figdir = os.path.join(outdir_figures)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c1428a86",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(outdir_figures):\n",
    "    os.makedirs(outdir_figures)\n",
    "if not os.path.exists(outdir_results):\n",
    "    os.makedirs(outdir_results)\n",
    "if not os.path.exists(outdir_export_data):\n",
    "    os.makedirs(outdir_export_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "28773239",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 322323 × 577\n",
       "    obs: 'CELL', 'CONDITION', 'BARCODE', 'ID', 'Donor', 'celltype0', 'celltype1', 'celltype2', 'dblabel', 'technology', 'tissue'\n",
       "    var: 'SYMBOL', 'ENSEMBL'\n",
       "    obsm: 'X_pca', 'X_umap'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Read annotated data\n",
    "adata_annotated = sc.read(input_annotated)\n",
    "adata_annotated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "64498a1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 753435 × 14966\n",
       "    obs: 'CELL', 'CONDITION', 'BARCODE', 'ID', 'Donor', 'technology', 'tissue'\n",
       "    var: 'SYMBOL', 'ENSEMBL'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read raw data\n",
    "adata_raw = sc.read(input_raw, citeseq = None)\n",
    "adata_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b193d102",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 322323 × 14966\n",
       "    obs: 'CELL', 'CONDITION', 'BARCODE', 'ID', 'Donor', 'celltype0', 'celltype1', 'celltype2', 'dblabel', 'technology', 'tissue'\n",
       "    var: 'SYMBOL', 'ENSEMBL'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = adata_raw[adata_annotated.obs_names].copy()\n",
    "adata.obs = adata_annotated.obs.copy()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "937bda4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14577.0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(adata.obs['celltype2'])\n",
    "\n",
    "adata.obs['ID'] = adata.obs['ID'].cat.reorder_categories([\n",
    "    'Donor1_PBMC_10Xfixed', \n",
    "    'Donor2_PBMC_10Xfixed', \n",
    "    'Donor3_PBMC_10Xfixed',\n",
    "    'Donor1_RBCdepl_10Xfixed', \n",
    "    'Donor2_RBCdepl_10Xfixed',\n",
    "    'Donor3_RBCdepl_10Xfixed', \n",
    "    'Donor1_PBMC_10X3prime',\n",
    "    'Donor2_PBMC_10X3prime', \n",
    "    'Donor3_PBMC_10X3prime',\n",
    "    'Donor1_RBCdepl_10X3prime', \n",
    "    'Donor2_RBCdepl_10X3prime',\n",
    "    'Donor3_RBCdepl_10X3prime', \n",
    "    'Donor1_RBCdepl_Parse',\n",
    "    'Donor3_RBCdepl_Parse', \n",
    "    'Donor2_RBCdepl_Parse', \n",
    "    'Donor1_PBMC_Parse', \n",
    "    'Donor2_PBMC_Parse',\n",
    "    'Donor3_PBMC_Parse',\n",
    "    'Donor1_WB_D0_HIVE',\n",
    "    'Donor2_WB_D0_HIVE',\n",
    "    'Donor3_WB_D0_HIVE',\n",
    "])\n",
    "adata.X.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4ad4d677",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/misaell2/.conda/envs/besca/lib/python3.9/site-packages/besca/st/_wrapper_funcs.py:533: FutureWarning:\n",
      "\n",
      "In the future, the default backend for leiden will be igraph instead of leidenalg.\n",
      "\n",
      " To achieve the future defaults please pass: flavor=\"igraph\" and n_iterations=2.  directed must also be False to work with igraph's implementation.\n",
      "\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'No \"neighbors\" in .uns'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m adata \u001b[38;5;241m=\u001b[39m \u001b[43mbc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mst\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclustering\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mresults_folder\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmyres\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1.0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m adata\n",
      "File \u001b[0;32m~/.conda/envs/besca/lib/python3.9/site-packages/besca/st/_wrapper_funcs.py:533\u001b[0m, in \u001b[0;36mclustering\u001b[0;34m(adata, results_folder, myres, method)\u001b[0m\n\u001b[1;32m    531\u001b[0m     sc_louvain(adata, resolution\u001b[38;5;241m=\u001b[39mmyres, random_state\u001b[38;5;241m=\u001b[39mrandom_state)\n\u001b[1;32m    532\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m method \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mleiden\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m--> 533\u001b[0m     \u001b[43msc_leiden\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mresolution\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmyres\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrandom_state\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrandom_state\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    534\u001b[0m \u001b[38;5;28mprint\u001b[39m(method \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m clustering performed with a resolution of \u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(myres))\n\u001b[1;32m    535\u001b[0m clusNum \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mset\u001b[39m(adata\u001b[38;5;241m.\u001b[39mobs[method]))\n",
      "File \u001b[0;32m~/.conda/envs/besca/lib/python3.9/site-packages/scanpy/tools/_leiden.py:150\u001b[0m, in \u001b[0;36mleiden\u001b[0;34m(adata, resolution, restrict_to, random_state, key_added, adjacency, directed, use_weights, n_iterations, partition_type, neighbors_key, obsp, copy, flavor, **clustering_args)\u001b[0m\n\u001b[1;32m    148\u001b[0m \u001b[38;5;66;03m# are we clustering a user-provided graph or the default AnnData one?\u001b[39;00m\n\u001b[1;32m    149\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m adjacency \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 150\u001b[0m     adjacency \u001b[38;5;241m=\u001b[39m \u001b[43m_utils\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_choose_graph\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mobsp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mneighbors_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    151\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m restrict_to \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    152\u001b[0m     restrict_key, restrict_categories \u001b[38;5;241m=\u001b[39m restrict_to\n",
      "File \u001b[0;32m~/.conda/envs/besca/lib/python3.9/site-packages/scanpy/_utils/__init__.py:1092\u001b[0m, in \u001b[0;36m_choose_graph\u001b[0;34m(adata, obsp, neighbors_key)\u001b[0m\n\u001b[1;32m   1090\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m adata\u001b[38;5;241m.\u001b[39mobsp[obsp]\n\u001b[1;32m   1091\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1092\u001b[0m     neighbors \u001b[38;5;241m=\u001b[39m \u001b[43mNeighborsView\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mneighbors_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1093\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mconnectivities\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m neighbors:\n\u001b[1;32m   1094\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1095\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mYou need to run `pp.neighbors` first \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1096\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mto compute a neighborhood graph.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1097\u001b[0m         )\n",
      "File \u001b[0;32m~/.conda/envs/besca/lib/python3.9/site-packages/scanpy/_utils/__init__.py:1023\u001b[0m, in \u001b[0;36mNeighborsView.__init__\u001b[0;34m(self, adata, key)\u001b[0m\n\u001b[1;32m   1021\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mor\u001b[39;00m key \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mneighbors\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m   1022\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mneighbors\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m adata\u001b[38;5;241m.\u001b[39muns:\n\u001b[0;32m-> 1023\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNo \u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mneighbors\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m in .uns\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m   1024\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_neighbors_dict \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39muns[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mneighbors\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1025\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_conns_key \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mconnectivities\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'No \"neighbors\" in .uns'"
     ]
    }
   ],
   "source": [
    "\n",
    "adata = bc.st.clustering(adata, results_folder, myres=1.0)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd298136",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, h5py, numpy as np, pandas as pd, scanpy as sc\n",
    "import scipy.sparse as sp\n",
    "from scipy.sparse import issparse\n",
    "\n",
    "# --- Paths (adjust if needed) ---\n",
    "input_annotated = os.path.join(os.getcwd(), 'annotated.all_merged.h5ad')\n",
    "input_raw       = os.path.join(os.getcwd(), 'raw.all_merged.h5ad')\n",
    "out_small       = 'raw.RBCsubset.h5ad'\n",
    "\n",
    "# 1) Load annotated (small-ish)\n",
    "adata_annot = sc.read_h5ad(input_annotated)\n",
    "\n",
    "# 2) RBC mask from annotated metadata\n",
    "rbc_mask = adata_annot.obs['ID'].astype(str).str.contains('RBC', case=False, na=False)\n",
    "rbc_indices = np.flatnonzero(rbc_mask.values)\n",
    "print(f\"RBC cells: {rbc_indices.size:,}\")\n",
    "\n",
    "# 3) Open raw .h5ad and stream only RBC rows from a CSR matrix\n",
    "with h5py.File(input_raw, \"r\") as f:\n",
    "    # --- get var names and shapes (using backed AnnData for reliability)\n",
    "    adata_raw_backed = sc.read_h5ad(input_raw, backed=\"r\")\n",
    "    var_names = adata_raw_backed.var_names.copy()\n",
    "    n_vars    = var_names.size\n",
    "    n_obs     = adata_raw_backed.n_obs\n",
    "    adata_raw_backed.file.close()   # important: close manually\n",
    "\n",
    "    # --- locate CSR storage for X\n",
    "    # anndata stores CSR arrays at: X/data, X/indices, X/indptr\n",
    "    Xgrp    = f['X']\n",
    "    indptr  = Xgrp['indptr']   # shape: (n_obs + 1,)\n",
    "    indices = Xgrp['indices']\n",
    "    data    = Xgrp['data']\n",
    "\n",
    "    # --- build CSR for the RBC subset by concatenating selected rows\n",
    "    out_data    = []\n",
    "    out_indices = []\n",
    "    out_indptr  = [0]\n",
    "\n",
    "    rbc_indices_sorted = np.sort(rbc_indices)\n",
    "    for i in rbc_indices_sorted:\n",
    "        start = int(indptr[i])\n",
    "        end   = int(indptr[i + 1])\n",
    "        if end > start:\n",
    "            out_data.append(   data[start:end]   )\n",
    "            out_indices.append(indices[start:end])\n",
    "        out_indptr.append(out_indptr[-1] + (end - start))\n",
    "\n",
    "    if out_data:\n",
    "        out_data    = np.concatenate(out_data)\n",
    "        out_indices = np.concatenate(out_indices)\n",
    "    else:\n",
    "        out_data    = np.array([], dtype=data.dtype)\n",
    "        out_indices = np.array([], dtype=indices.dtype)\n",
    "\n",
    "# 4) Construct smaller AnnData for RBC cells\n",
    "X_rbc = sp.csr_matrix((out_data, out_indices, np.array(out_indptr, dtype=np.int64)),\n",
    "                      shape=(rbc_indices.size, n_vars))\n",
    "\n",
    "obs_rbc = adata_annot.obs.iloc[rbc_indices].copy()\n",
    "var_df  = pd.DataFrame(index=var_names)  # minimal var frame\n",
    "\n",
    "adata_rbc = sc.AnnData(X=X_rbc, obs=obs_rbc, var=var_df)\n",
    "\n",
    "# 5) Precompute QC columns on the small object (sparse-safe)\n",
    "def row_sum(mat):\n",
    "    return np.asarray(mat.sum(axis=1)).ravel() if issparse(mat) else mat.sum(axis=1)\n",
    "\n",
    "def row_nnz(mat):\n",
    "    if issparse(mat):\n",
    "        # (mat > 0) is OK for CSR; avoid densifying\n",
    "        return np.asarray((mat > 0).sum(axis=1)).ravel()\n",
    "    return (mat > 0).sum(axis=1)\n",
    "\n",
    "# total UMI and detected genes\n",
    "umi_counts  = row_sum(adata_rbc.X).astype(np.float64)\n",
    "gene_counts = row_nnz(adata_rbc.X).astype(np.int64)\n",
    "adata_rbc.obs['total_counts']       = umi_counts\n",
    "adata_rbc.obs['n_genes_by_counts']  = gene_counts\n",
    "\n",
    "# mitochondrial fraction (handles MT- or MT_)\n",
    "genes  = np.asarray(adata_rbc.var_names, dtype=str)\n",
    "upper  = np.char.upper(genes)\n",
    "mt_mask = np.char.startswith(upper, 'MT-') | np.char.startswith(upper, 'MT_')\n",
    "\n",
    "if mt_mask.any():\n",
    "    mt_counts = row_sum(adata_rbc.X[:, mt_mask]).astype(np.float64)\n",
    "else:\n",
    "    mt_counts = np.zeros(adata_rbc.n_obs, dtype=np.float64)\n",
    "\n",
    "umi_safe = umi_counts.copy()\n",
    "umi_safe[umi_safe == 0] = np.nan\n",
    "adata_rbc.obs['mito_fraction'] = mt_counts / umi_safe\n",
    "\n",
    "# 6) Save the manageable file for downstream plotting\n",
    "adata_rbc.write(out_small, compression='gzip')\n",
    "print(f\"Wrote {out_small} with shape {adata_rbc.shape}\")\n",
    "print(adata_rbc.obs[['total_counts','n_genes_by_counts','mito_fraction']].describe().T)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84c8e0d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "rbc_cells = sc.read_h5ad(\"raw.RBCsubset.h5ad\")\n",
    "print(rbc_cells)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "985c9eb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "# map variables to right columns\n",
    "tech_map = {\n",
    "    '10X3prime': 'Chromium Single Cell 3′ Gene Expression v3.1 (10X Genomics)',\n",
    "    '10Xfixed':  'Chromium Single Cell Gene Expression Flex (10X Genomics)',\n",
    "    'Parse':     'Evercode WT Mini v2 (Parse Biosciences)',\n",
    "    'HIVE':      'HIVE scRNAseq v1 (Honeycomb Biotechnologies)',\n",
    "}\n",
    "\n",
    "\n",
    "# Prepare tidy DF\n",
    "df_qc = pd.DataFrame({\n",
    "    \"Donor\": rbc_cells.obs[\"Donor\"].astype(str),\n",
    "    \"Technology_code\": rbc_cells.obs[\"technology\"].astype(str),\n",
    "    \"UMI counts\": rbc_cells.obs[\"total_counts\"].values,\n",
    "    \"Mitochondrial fraction\": rbc_cells.obs[\"mito_fraction\"].values,\n",
    "})\n",
    "\n",
    "# Order the technologies and donors\n",
    "tech_order = [t for t in [\"10X3prime\", \"10Xfixed\", \"Parse\", \"HIVE\"]\n",
    "              if t in df_qc[\"Technology_code\"].unique()]\n",
    "\n",
    "donor_order = sorted(df_qc[\"Donor\"].unique().tolist())\n",
    "\n",
    "# Color map (tech → color)\n",
    "palette = sns.color_palette(\"Set2\", n_colors=len(tech_order))\n",
    "tech_color_map = {tech_order[i]: palette[i] for i in range(len(tech_order))}\n",
    "\n",
    "# Faceted violin plot: one column per technology\n",
    "g = sns.catplot(\n",
    "    data=df_qc,\n",
    "    kind=\"violin\",\n",
    "    col=\"Technology_code\",\n",
    "    x=\"Donor\",\n",
    "    y=\"Mitochondrial fraction\",\n",
    "    order=donor_order,\n",
    "    col_order=tech_order,\n",
    "    hue=\"Technology_code\",\n",
    "    palette=tech_color_map,\n",
    "    inner=\"quartile\",\n",
    "    cut=0,\n",
    "    sharey=True,\n",
    "    height=4,\n",
    "    aspect=1.1\n",
    ")\n",
    "\n",
    "# y-range to match the paper\n",
    "g.set(ylim=(0, 0.25))\n",
    "\n",
    "# Replace top titles with full descriptive labels\n",
    "for ax, tech in zip(g.axes.flatten(), tech_order):\n",
    "    ax.set_title(tech_map.get(tech, tech))\n",
    "\n",
    "# Clean x-axis\n",
    "g.set_xlabels(\"Donor\")\n",
    "g.set_ylabels(\"Mitochondrial fraction\")\n",
    "\n",
    "# Remove the legend (color is redundant per panel)\n",
    "g._legend.remove()\n",
    "\n",
    "plt.suptitle(\"Figure 2A: Mitochondrial fraction per Donor within each Technology\", y=1.05)\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d9d6a42",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c5033b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.catplot(\n",
    "    data=df_qc,\n",
    "    kind=\"violin\",\n",
    "    col=\"Technology_code\",\n",
    "    x=\"Donor\",\n",
    "    y=\"UMI counts\",\n",
    "    order=donor_order,\n",
    "    col_order=tech_order,\n",
    "    hue=\"Technology_code\",\n",
    "    palette=tech_color_map,\n",
    "    inner=\"quartile\",\n",
    "    cut=0,\n",
    "    sharey=True,\n",
    "    height=4,\n",
    "    aspect=1.1\n",
    ")\n",
    "\n",
    "g.set(ylim=(0, 25000))\n",
    "\n",
    "# Replace top titles\n",
    "for ax, tech in zip(g.axes.flatten(), tech_order):\n",
    "    ax.set_title(tech_map.get(tech, tech))\n",
    "\n",
    "g.set_xlabels(\"Donor\")\n",
    "g.set_ylabels(\"UMI counts\")\n",
    "\n",
    "# Remove legend — redundant because each panel is a single tech\n",
    "g._legend.remove()\n",
    "\n",
    "plt.suptitle(\"Figure 2B: UMI counts per Donor within each Technology\", y=1.05)\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "035b2cab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# =============================================================================\n",
    "# Figure 2C — “Number of genes” per Donor × Technology (RBC-depleted cells)\n",
    "# =============================================================================\n",
    "# Goal:\n",
    "#   • Reproduce Figure 2C from the paper, using the *n_genes_by_counts*\n",
    "#     QC metric as the y-axis.\n",
    "#   • Each panel (column) represents a Technology (10X 3′, 10X Flex, Parse, HIVE).\n",
    "#   • Within each technology, the x-axis shows Donor 1–3 side by side.\n",
    "#   • Violin colors correspond to technology; all violins share the same y-range.\n",
    "#\n",
    "# Inputs:\n",
    "#   rbc_cells.obs['Donor']               – donor/sample ID per cell\n",
    "#   rbc_cells.obs['technology']          – short technology codes (10X3prime, 10Xfixed, Parse, HIVE)\n",
    "#   rbc_cells.obs['n_genes_by_counts']   – number of detected genes per cell\n",
    "#\n",
    "# Expected output:\n",
    "#   A multi-panel figure (one column per technology), each showing three violins\n",
    "#   (Donor 1–3) colored by technology, y-axis from 0 → 5000.\n",
    "\n",
    "# -------------------------------------------------------------------------\n",
    "# 1) Build a tidy DataFrame for plotting\n",
    "# -------------------------------------------------------------------------\n",
    "df_qc_genes = pd.DataFrame({\n",
    "    \"Donor\": rbc_cells.obs[\"Donor\"].astype(str),\n",
    "    \"Technology_code\": rbc_cells.obs[\"technology\"].astype(str),\n",
    "    \"Number of genes\": rbc_cells.obs[\"n_genes_by_counts\"].astype(float),\n",
    "})\n",
    "\n",
    "# Order technologies (only those present in the current dataset)\n",
    "tech_order = [t for t in [\"10X3prime\", \"10Xfixed\", \"Parse\", \"HIVE\"]\n",
    "              if t in df_qc_genes[\"Technology_code\"].unique()]\n",
    "\n",
    "# Order donors alphabetically/numerically for consistent x-axis placement\n",
    "donor_order = sorted(df_qc_genes[\"Donor\"].unique().tolist())\n",
    "\n",
    "# -------------------------------------------------------------------------\n",
    "# 2) Define color palette (consistent with earlier figures)\n",
    "# -------------------------------------------------------------------------\n",
    "# We assign one distinct color per technology for visual consistency\n",
    "base_palette = sns.color_palette(\"Set2\", n_colors=4)\n",
    "tech_color_map = {code: base_palette[i] for i, code in enumerate([\"10X3prime\", \"10Xfixed\", \"Parse\", \"HIVE\"])}\n",
    "palette_list = [tech_color_map[c] for c in tech_order]\n",
    "\n",
    "# -------------------------------------------------------------------------\n",
    "# 3) Plot — one panel per technology, violins grouped by donor\n",
    "# -------------------------------------------------------------------------\n",
    "g = sns.catplot(\n",
    "    data=df_qc_genes,\n",
    "    kind=\"violin\",\n",
    "    col=\"Technology_code\",        # → each column = one technology\n",
    "    x=\"Donor\",                    # → donors on x-axis within each technology\n",
    "    y=\"Number of genes\",          # → QC metric on y-axis\n",
    "    order=donor_order,            # consistent donor order\n",
    "    col_order=tech_order,         # consistent tech order (left→right)\n",
    "    hue=\"Technology_code\",        # color violins by technology\n",
    "    palette=tech_color_map,       # fixed color map (ensures consistency)\n",
    "    inner=\"quartile\",             # show median + quartile lines inside violins\n",
    "    cut=0,                        # don't extend violins beyond data range\n",
    "    sharey=True,                  # same y-axis scale across panels\n",
    "    height=4,                     # panel height (in inches)\n",
    "    aspect=1.1                    # width/height ratio\n",
    ")\n",
    "\n",
    "# -------------------------------------------------------------------------\n",
    "# 4) Styling / axes / titles\n",
    "# -------------------------------------------------------------------------\n",
    "# Set a consistent y-axis range to match the paper\n",
    "g.set(ylim=(0, 5000))\n",
    "\n",
    "# Replace the short technology codes in the titles with descriptive names\n",
    "tech_map = {\n",
    "    \"10X3prime\": \"Chromium Single Cell 3′ Gene Expression v3.1 (10X Genomics)\",\n",
    "    \"10Xfixed\":  \"Chromium Single Cell Gene Expression Flex (10X Genomics)\",\n",
    "    \"Parse\":     \"Evercode WT Mini v2 (Parse Biosciences)\",\n",
    "    \"HIVE\":      \"HIVE scRNAseq v1 (Honeycomb Biotechnologies)\",\n",
    "}\n",
    "for ax, tech in zip(g.axes.flatten(), tech_order):\n",
    "    ax.set_title(tech_map.get(tech, tech))\n",
    "\n",
    "# Label axes cleanly\n",
    "g.set_xlabels(\"Donor\")\n",
    "g.set_ylabels(\"Number of genes detected per cell\")\n",
    "\n",
    "# Remove legend (redundant, color = panel)\n",
    "if g._legend is not None:\n",
    "    g._legend.remove()\n",
    "\n",
    "# Global figure title and layout\n",
    "plt.suptitle(\"Figure 2C – Number of genes per Donor within each Technology (RBC-depleted)\", y=1.05)\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# -------------------------------------------------------------------------\n",
    "# 5) Sanity checks / interpretation\n",
    "# -------------------------------------------------------------------------\n",
    "# • Each technology panel shows 3 violins (Donor 1–3).\n",
    "# • The y-axis range (0–5000) highlights that typical single-cell libraries\n",
    "#   detect hundreds to a few thousand genes per cell.\n",
    "# • Expect Parse and 10X Flex to cluster near 1–3 k genes; 10X 3′ may extend higher.\n",
    "# • If a panel is missing → that technology had no RBC-depleted cells in your dataset.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5b0ab0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# =============================================================================\n",
    "# Figure 2D (extended) — Dot plot across Technologies *and* Cell Types\n",
    "# =============================================================================\n",
    "# Goal:\n",
    "#   • Rows = first each Technology, then key immune Cell Types (B/T/Mono/NK/Neutrophil).\n",
    "#   • Columns = selected genes (stable markers).\n",
    "#   • Dot size  = fraction of cells in the group expressing the gene (>0 counts).\n",
    "#   • Dot color = log1p(mean raw expression) in the group (white→dark red).\n",
    "#\n",
    "# Inputs (from previous steps):\n",
    "#   rbc_cells: AnnData with raw counts in .X and metadata in .obs\n",
    "#   rbc_cells.obs['technology']: technology code ('10X3prime','10Xfixed','Parse','HIVE')\n",
    "#   rbc_cells.obs['celltype*'] : cell type labels (we'll auto-pick a suitable column)\n",
    "#   rbc_cells.var_names: gene symbols\n",
    "#\n",
    "# Output:\n",
    "#   A composite dot plot with Technologies on top followed by selected Cell Types;\n",
    "#   right side shows a size legend for fraction (%) and a colorbar for mean expression.\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.sparse import issparse\n",
    "\n",
    "# -----------------------------\n",
    "# 1) Configure genes & columns\n",
    "# -----------------------------\n",
    "genes_for_dotplot = [\n",
    "    \"S100A8\", \"S100A9\", \"MPO\", \"ELANE\", \"LTF\", \"CEACAM8\", \"CXCR2\",\n",
    "    \"CSF3R\", \"FCGR3B\", \"OLFM4\", \"LYZ\", \"GAPDH\"\n",
    "]\n",
    "present_genes = [g for g in genes_for_dotplot if g in rbc_cells.var_names]\n",
    "if not present_genes:\n",
    "    raise ValueError(\"None of the selected genes are present in rbc_cells.var_names.\")\n",
    "\n",
    "# Pick a suitable cell-type column\n",
    "celltype_candidates = [\"celltype1\", \"celltype0\", \"celltype2\", \"cell_type\", \"CellType\"]\n",
    "celltype_col = next((c for c in celltype_candidates if c in rbc_cells.obs.columns), None)\n",
    "if celltype_col is None:\n",
    "    raise KeyError(\"Could not find a cell-type column in rbc_cells.obs.\")\n",
    "\n",
    "# Cell types to include (in the paper’s order)\n",
    "wanted_celltypes = [\"B cell\", \"T cell\", \"monocyte\", \"natural killer cell\", \"neutrophil\"]\n",
    "\n",
    "# Technologies (order for rows)\n",
    "tech_codes = rbc_cells.obs[\"technology\"].astype(str).unique().tolist()\n",
    "tech_order = [t for t in [\"10X3prime\", \"10Xfixed\", \"Parse\", \"HIVE\"] if t in tech_codes]\n",
    "\n",
    "# Pretty mapping for technology labels (used in y-tick text)\n",
    "tech_map = {\n",
    "    \"10X3prime\": \"Chromium Single Cell 3′ Gene Expression v3.1 (10X Genomics)\",\n",
    "    \"10Xfixed\":  \"Chromium Single Cell Gene Expression Flex (10X Genomics)\",\n",
    "    \"Parse\":     \"Evercode WT Mini v2 (Parse Biosciences)\",\n",
    "    \"HIVE\":      \"HIVE scRNAseq v1 (Honeycomb Biotechnologies)\",\n",
    "}\n",
    "\n",
    "# Helper: dense view\n",
    "def to_dense(m):\n",
    "    return m.toarray() if issparse(m) else np.asarray(m)\n",
    "\n",
    "# ---------------------------------------------\n",
    "# 2) Aggregate per-group (tech and cell type)\n",
    "# ---------------------------------------------\n",
    "rows = []\n",
    "\n",
    "# Per-technology\n",
    "for tech in tech_order:\n",
    "    sub = rbc_cells[rbc_cells.obs[\"technology\"].astype(str) == tech]\n",
    "    X = to_dense(sub[:, present_genes].X)\n",
    "    frac = (X > 0).mean(axis=0)        # fraction expressing\n",
    "    mean_expr = X.mean(axis=0)         # mean raw counts\n",
    "    for g, f, m in zip(present_genes, frac, mean_expr):\n",
    "        rows.append({\n",
    "            \"Group\": tech,             # keep as the code; map later for y-axis text\n",
    "            \"GroupType\": \"Technology\",\n",
    "            \"Gene\": g,\n",
    "            \"Fraction\": f,\n",
    "            \"MeanExpr\": np.log1p(m)    # log1p scaling for color\n",
    "        })\n",
    "\n",
    "# Per-cell-type (across ALL technologies)\n",
    "present_ct = [ct for ct in wanted_celltypes if ct in set(rbc_cells.obs[celltype_col].astype(str))]\n",
    "for ct in present_ct:\n",
    "    sub = rbc_cells[rbc_cells.obs[celltype_col].astype(str) == ct]\n",
    "    X = to_dense(sub[:, present_genes].X)\n",
    "    frac = (X > 0).mean(axis=0)\n",
    "    mean_expr = X.mean(axis=0)\n",
    "    for g, f, m in zip(present_genes, frac, mean_expr):\n",
    "        rows.append({\n",
    "            \"Group\": ct,               # exact cell type string; no mapping\n",
    "            \"GroupType\": \"CellType\",\n",
    "            \"Gene\": g,\n",
    "            \"Fraction\": f,\n",
    "            \"MeanExpr\": np.log1p(m)\n",
    "        })\n",
    "\n",
    "dot_df = pd.DataFrame(rows)\n",
    "\n",
    "# ---------------------------------------------\n",
    "# 3) Build y-axis order: techs, then cell types\n",
    "# ---------------------------------------------\n",
    "y_order = tech_order + present_ct\n",
    "# Map tech codes → descriptive labels for y-ticks; keep cell-type names as-is\n",
    "def pretty_group_label(x):\n",
    "    return tech_map.get(x, x)\n",
    "\n",
    "# Ensure categorical ordering\n",
    "dot_df[\"Group\"] = pd.Categorical(dot_df[\"Group\"], categories=y_order, ordered=True)\n",
    "dot_df[\"Gene\"] = pd.Categorical(dot_df[\"Gene\"], categories=present_genes, ordered=True)\n",
    "\n",
    "# ---------------------------------------------\n",
    "# 4) Scatter plot: size=fraction, color=log1p(mean)\n",
    "# ---------------------------------------------\n",
    "fig_h = 0.8 * len(y_order) + 2\n",
    "fig_w = 1.1 * len(present_genes) + 3\n",
    "fig, ax = plt.subplots(figsize=(fig_w, fig_h))\n",
    "\n",
    "# Coordinates on grid\n",
    "x_pos = {g: i for i, g in enumerate(present_genes)}\n",
    "y_pos = {grp: i for i, grp in enumerate(y_order)}\n",
    "\n",
    "xs = dot_df[\"Gene\"].map(x_pos).values\n",
    "ys = dot_df[\"Group\"].map(y_pos).values\n",
    "\n",
    "# Fraction → size (% legend will show 20–100)\n",
    "sizes = (dot_df[\"Fraction\"].values * 100.0)  # as percent\n",
    "size_scale = 9.0  # visual multiplier (tweak if dots look too small/large)\n",
    "sizes_pts = (sizes * size_scale)            # points^2 for scatter\n",
    "\n",
    "# Color = log1p(mean expression); use white→dark red\n",
    "colors = dot_df[\"MeanExpr\"].values\n",
    "sc = ax.scatter(xs, ys, s=sizes_pts, c=colors, cmap=\"Reds\", edgecolor=\"none\")\n",
    "\n",
    "# Axes\n",
    "ax.set_xticks(range(len(present_genes)))\n",
    "ax.set_xticklabels(present_genes, rotation=45, ha=\"right\")\n",
    "ax.set_yticks(range(len(y_order)))\n",
    "ax.set_yticklabels([pretty_group_label(y) for y in y_order])\n",
    "\n",
    "ax.set_xlabel(\"Gene\")\n",
    "ax.set_ylabel(\"Group (Technology then Cell type)\")\n",
    "ax.set_title(\"Figure 2D — Fraction of cells (size) and mean expression (color) across groups\")\n",
    "\n",
    "# ---------------------------------------------\n",
    "# 5) Legends on the right\n",
    "# ---------------------------------------------\n",
    "# (A) Size legend: Fraction (%) = 20, 40, 60, 80, 100\n",
    "legend_percents = [20, 40, 60, 80, 100]\n",
    "legend_sizes = [p * size_scale for p in legend_percents]\n",
    "\n",
    "# Dummy handles for size legend\n",
    "handles = [plt.scatter([], [], s=s, color=\"gray\", edgecolor=\"none\") for s in legend_sizes]\n",
    "labels = [f\"{p}\" for p in legend_percents]\n",
    "\n",
    "size_legend = ax.legend(\n",
    "    handles, labels,\n",
    "    title=\"Fraction of cells in group (%)\",\n",
    "    loc=\"center left\",\n",
    "    bbox_to_anchor=(1.02, 0.5),\n",
    "    frameon=True\n",
    ")\n",
    "ax.add_artist(size_legend)\n",
    "\n",
    "# (B) Colorbar: mean expression in group (log1p scale with ticks 1,2,3)\n",
    "cbar = fig.colorbar(sc, ax=ax, fraction=0.046, pad=0.04)\n",
    "cbar.set_label(\"mean expression in group\")\n",
    "cbar.set_ticks([1, 2, 3])\n",
    "cbar.set_ticklabels([\"1\", \"2\", \"3\"])\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# ---------------------------------------------\n",
    "# 6) Notes\n",
    "# ---------------------------------------------\n",
    "# • Rows list all technologies first (pretty mapped names), then five immune cell types.\n",
    "# • Dot size legend shows fraction (%) with reference sizes at 20/40/60/80/100.\n",
    "# • Colorbar uses log1p(mean) but is *labeled* with 1/2/3 for easy reading.\n",
    "# • If a row is missing, that group had no RBC-depleted cells (or cell type not present).\n",
    "# • Adjust `size_scale` if dots look too large/small for your monitor or export size.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb422ca9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# =============================================================================\n",
    "# Figure 3A — Same UMAP embedding, colored by Technology / Tissue / Cell type\n",
    "# =============================================================================\n",
    "# Inputs:\n",
    "#   adata_annot: annotated AnnData with metadata in .obs and UMAP in .obsm['X_umap']\n",
    "# Required .obs columns:\n",
    "#   - 'technology' (codes like 10X3prime, 10Xfixed, Parse, HIVE)\n",
    "#   - 'tissue' (PBMC; RBC-depleted may be embedded in 'ID' field)\n",
    "#   - one of ['celltype1','celltype0','celltype2','dblabel'] for cell identity\n",
    "#\n",
    "# Output:\n",
    "#   Three panels with identical UMAP shapes:\n",
    "#     Left:  colored by technology\n",
    "#     Middle: colored by tissue (PBMC vs RBCdepl)\n",
    "#     Right: colored by cell type\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# ----------------------------\n",
    "# 1) Build a plotting DataFrame\n",
    "# ----------------------------\n",
    "if 'X_umap' not in adata_annot.obsm_keys():\n",
    "    raise KeyError(\"UMAP coordinates not found in adata_annot.obsm['X_umap'].\")\n",
    "\n",
    "umap = adata_annot.obsm['X_umap']\n",
    "df = pd.DataFrame({\n",
    "    'UMAP1': umap[:, 0],\n",
    "    'UMAP2': umap[:, 1],\n",
    "})\n",
    "\n",
    "# Technology codes (required)\n",
    "if 'technology' not in adata_annot.obs.columns:\n",
    "    raise KeyError(\"Expected 'technology' in adata_annot.obs.\")\n",
    "df['technology'] = adata_annot.obs['technology'].astype(str).values\n",
    "\n",
    "# Tissue:\n",
    "# - Use 'tissue' if present (often \"PBMC\")\n",
    "# - Add \"RBCdepl\" based on 'ID' substring (case-insensitive)\n",
    "tissue_series = None\n",
    "if 'tissue' in adata_annot.obs.columns:\n",
    "    tissue_series = adata_annot.obs['tissue'].astype(str)\n",
    "else:\n",
    "    tissue_series = pd.Series(['Unknown'] * adata_annot.n_obs, index=adata_annot.obs_names)\n",
    "\n",
    "if 'ID' in adata_annot.obs.columns:\n",
    "    id_str = adata_annot.obs['ID'].astype(str)\n",
    "    is_rbc = id_str.str.contains('RBC', case=False, na=False)\n",
    "    # Prefer explicit RBCdepl over generic tissue label\n",
    "    tissue_series = np.where(is_rbc, 'RBCdepl', tissue_series)\n",
    "\n",
    "df['tissue'] = pd.Series(tissue_series, index=adata_annot.obs_names).astype(str).values\n",
    "\n",
    "# Cell type column: pick the best available\n",
    "celltype_candidates = ['celltype1','celltype0','celltype2','dblabel']\n",
    "celltype_col = next((c for c in celltype_candidates if c in adata_annot.obs.columns), None)\n",
    "if celltype_col is None:\n",
    "    raise KeyError(\"No cell type column found (looked for 'celltype1','celltype0','celltype2','dblabel').\")\n",
    "\n",
    "df['celltype'] = adata_annot.obs[celltype_col].astype(str).values\n",
    "\n",
    "# ---------------------------------------\n",
    "# 2) Category orders and nice label maps\n",
    "# ---------------------------------------\n",
    "tech_order = [t for t in ['10X3prime','10Xfixed','Parse','HIVE'] if t in df['technology'].unique()]\n",
    "tech_map = {\n",
    "    '10X3prime': \"Chromium Single Cell 3′ Gene Expression v3.1 (10X Genomics)\",\n",
    "    '10Xfixed':  \"Chromium Single Cell Gene Expression Flex (10X Genomics)\",\n",
    "    'Parse':     \"Evercode WT Mini v2 (Parse Biosciences)\",\n",
    "    'HIVE':      \"HIVE scRNAseq v1 (Honeycomb Biotechnologies)\",\n",
    "}\n",
    "tech_palette = sns.color_palette('Set2', n_colors=max(1, len(tech_order)))\n",
    "tech_color_map = {code: tech_palette[i] for i, code in enumerate(tech_order)}\n",
    "\n",
    "# Tissue order & colors (PBMC vs RBCdepl)\n",
    "tissue_order = [t for t in ['PBMC','RBCdepl'] if t in df['tissue'].unique()]\n",
    "# Fallback: keep whatever is there, but ensure deterministic order\n",
    "if not tissue_order:\n",
    "    tissue_order = sorted(df['tissue'].unique().tolist())\n",
    "\n",
    "tissue_color_map = {\n",
    "    'PBMC':    '#1f77b4',  # blue\n",
    "    'RBCdepl': '#d62728',  # red\n",
    "}\n",
    "# Fill any unknowns with gray\n",
    "for t in df['tissue'].unique():\n",
    "    tissue_color_map.setdefault(t, '#7f7f7f')\n",
    "\n",
    "# Cell type order: put common immune types first if present\n",
    "preferred_celltypes = [\n",
    "    'B cell', 'T cell', 'Natural killer cell', 'monocyte',\n",
    "    'neutrophil', 'granulocyte', 'Hematopoietic multipotent progenitor cell'\n",
    "]\n",
    "present_celltypes = df['celltype'].unique().tolist()\n",
    "celltype_order = [c for c in preferred_celltypes if c in present_celltypes]\n",
    "celltype_order += [c for c in present_celltypes if c not in celltype_order]  # append the rest\n",
    "\n",
    "# Distinct palette for cell types (extend as needed)\n",
    "celltype_colors = sns.color_palette('tab20', n_colors=len(celltype_order))\n",
    "celltype_color_map = {c: celltype_colors[i] for i, c in enumerate(celltype_order)}\n",
    "\n",
    "# ---------------------------------------\n",
    "# 3) Plot — three panels, same axes limits\n",
    "# ---------------------------------------\n",
    "# Compute shared limits so shapes are identical\n",
    "xlim = (df['UMAP1'].min() - 0.5, df['UMAP1'].max() + 0.5)\n",
    "ylim = (df['UMAP2'].min() - 0.5, df['UMAP2'].max() + 0.5)\n",
    "\n",
    "fig, axes = plt.subplots(1, 3, figsize=(18, 5), constrained_layout=True)\n",
    "\n",
    "# Panel 1: Technology\n",
    "ax = axes[0]\n",
    "# map colors for each point\n",
    "colors_tech = df['technology'].map(lambda t: tech_color_map.get(t, '#7f7f7f'))\n",
    "ax.scatter(df['UMAP1'], df['UMAP2'], c=colors_tech, s=4, alpha=0.7, linewidths=0)\n",
    "ax.set_title('Technology')\n",
    "ax.set_xlim(*xlim); ax.set_ylim(*ylim)\n",
    "ax.set_xticks([]); ax.set_yticks([])\n",
    "# Legend (compact)\n",
    "handles = [plt.Line2D([0],[0], marker='o', color='w', markerfacecolor=tech_color_map[c], markersize=8, label=tech_map.get(c,c))\n",
    "           for c in tech_order]\n",
    "ax.legend(handles=handles, title='', loc='lower left', fontsize=8, frameon=False)\n",
    "\n",
    "# Panel 2: Tissue (PBMC vs RBCdepl)\n",
    "ax = axes[1]\n",
    "colors_tissue = df['tissue'].map(lambda t: tissue_color_map.get(t, '#7f7f7f'))\n",
    "ax.scatter(df['UMAP1'], df['UMAP2'], c=colors_tissue, s=4, alpha=0.7, linewidths=0)\n",
    "ax.set_title('Tissue')\n",
    "ax.set_xlim(*xlim); ax.set_ylim(*ylim)\n",
    "ax.set_xticks([]); ax.set_yticks([])\n",
    "handles = [plt.Line2D([0],[0], marker='o', color='w', markerfacecolor=tissue_color_map[c], markersize=8, label=c)\n",
    "           for c in tissue_order]\n",
    "ax.legend(handles=handles, title='', loc='lower left', fontsize=8, frameon=False)\n",
    "\n",
    "# Panel 3: Cell type\n",
    "ax = axes[2]\n",
    "colors_cell = df['celltype'].map(lambda c: celltype_color_map.get(c, '#7f7f7f'))\n",
    "ax.scatter(df['UMAP1'], df['UMAP2'], c=colors_cell, s=4, alpha=0.7, linewidths=0)\n",
    "ax.set_title('Cell type')\n",
    "ax.set_xlim(*xlim); ax.set_ylim(*ylim)\n",
    "ax.set_xticks([]); ax.set_yticks([])\n",
    "# Put a scrollable-feel legend (may be long)\n",
    "handles = [plt.Line2D([0],[0], marker='o', color='w', markerfacecolor=celltype_color_map[c], markersize=6, label=c)\n",
    "           for c in celltype_order[:15]]  # show top ~15 to avoid clutter; adjust as needed\n",
    "ax.legend(handles=handles, title='', loc='lower left', fontsize=7, frameon=False)\n",
    "\n",
    "plt.suptitle('Figure 3A — Same UMAP, colored by Technology / Tissue / Cell type', y=1.02)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a9d2bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = bc.st.clustering(adata, results_folder, myres=1.0)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e45c4a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "if rbc_cells.obs.get(\"percent_mito\") is None:\n",
    "    bc.pp.fraction_counts(adata=adata)\n",
    "if adata.obs.get(\"n_counts\") is None:\n",
    "    adata.obs[\"n_counts\"] = adata.X.sum(axis=1)\n",
    "if adata.obs.get(\"n_genes\") is None:\n",
    "    adata.obs[\"n_genes\"] = np.sum(adata.X > 0, axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f5170c9",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "besca",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
